### PR TITLE
Header Footer: move from calypso to package

### PR DESCRIPTION
**Pull request by arcangelini**
## Proposed Changes

With the recent projects to unify our WordPress.com sites such as /support, /forums, and /learn we are lacking a unified approach to loading the constants like the header and footer. Both of these components have existed inside Calypso for the logged-out view of themes and plugins.

This PR moves the footer and header out of `client/layout` to `packages/wpcom-template-parts`. The goal from after this is to use them inside a block found in `apps/happy-blocks`. This PR also moves the WordPress watermark to the components package as well.

Markup 2023-02-20 at 17 44 41

Markup 2023-02-20 at 17 45 01

## Testing Instructions

1. Pull branch and `yarn start`
2. Go to the two sites that these components are used: http://calypso.localhost:3000/plugins & http://calypso.localhost:3000/themes
3. Compare against https://wordpress.com/plugins & https://wordpress.com/themes
4. Test language switcher and links.
5. I am not exactly share what conditions need to be met for DoNotSellLink to be true, but you can change to be true here: https://github.com/Automattic/wp-calypso/blob/9ac12871be9ba149bf1e4a75a2cd7433cb387d76/client/layout/logged-out.jsx#L46

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

**Comment by arcangelini** [comment]
Not sure exactly why, but my PR was complaining about a few random errors so I chose to resolve them to the best of my ability. Tagging the last people to work in them to make sure I didn't break anything or for better input 😄 

`packages/page-pattern-modal/src/utils/replace-placeholders.ts` - @p-jackson 
`packages/components/src/list-tile/index.tsx` - @zaguiini 
`packages/calypso-e2e/src/rest-api-client.ts` - @worldomonation (This one was tricky because fixing it caused an error to throw in the API response)

**Comment by zaguiini** [comment]
Good to go from my end @arcangelini -- it's just some typing imports rewrite, nothing breaking 👍

File: packages/calypso-e2e/src/rest-api-client.ts [hunk 1]
```diff
@@ -358,7 +358,7 @@ export class RestAPIClient {
 			);
 		}
 
-		if ( response.errors === [] ) {
+		if ( response.errors.length > 0 && ! response.sent ) {
```
Comment (worldomonation) [inline packages/calypso-e2e/src/rest-api-client.ts:361]: **question:blocking** - reading the description of this PR it is unexpected to see a change here. Was it failing tests?

File: packages/page-pattern-modal/src/utils/replace-placeholders.ts [hunk 2]
```diff
@@ -32,7 +32,7 @@ const replacePlaceholders = (
 };
 
 // eslint-disable-next-line @typescript-eslint/no-explicit-any
-function isObjKey< T >( key: any, obj: T ): key is keyof T {
+function isObjKey< T >( key: any, obj: any ): key is keyof T {
```
Comment (p-jackson) [inline packages/page-pattern-modal/src/utils/replace-placeholders.ts]: Thanks for the ping @arcangelini. Unfortunately changing the type of `obj` to `any` means that type `T` no longer gets inferred and the point of the function stops working. What I should have written in the first place is this:

```suggestion
function isObjKey< T extends object >( key: any, obj: T ): key is keyof T {
```

so that TypeScript knows we're allowed to use the `in` keyword on the next line.

File: client/layout/logged-out.jsx [hunk 3]
```diff
@@ -29,6 +31,34 @@ import BodySectionCssClass from './body-section-css-class';
 
 import './style.scss';
 
+const DoNotSellLink = () => {
```
Comment (tyxla) [inline client/layout/logged-out.jsx:34]: Is this component essential? It seems like something that can be async-loaded.

File: packages/calypso-e2e/src/rest-api-client.ts (hunk 1 above)
Comment (tyxla) [reply packages/calypso-e2e/src/rest-api-client.ts:361]: Agreed with @worldomonation. It seems like a change that can (and should) be moved to another PR.

File: packages/components/src/list-tile/index.tsx [hunk 4]
```diff
@@ -1,13 +1,13 @@
 /* eslint-disable no-nested-ternary */
 import classNames from 'classnames';
-import React, { cloneElement, isValidElement } from 'react';
+import { cloneElement, isValidElement, ReactElement, ReactNode } from 'react';
```
Comment (tyxla) [inline packages/components/src/list-tile/index.tsx:3]: Should we import types with `import type`?

File: packages/components/src/site-thumbnail/docs/example.jsx [hunk 5]
```diff
@@ -1,4 +1,4 @@
-import WordPressLogo from '../../wordpress-logo';
+import { WordPressLogo } from '../../wordpress-logo';
```
Comment (tyxla) [inline packages/components/src/site-thumbnail/docs/example.jsx:1]: Those changes are yet another example of something that can be moved to another PR.

File: packages/components/src/wordpress-logo/docs/example.jsx [hunk 6]
```diff
@@ -1,4 +1,4 @@
-import WordPressLogo from '../index';
+import { WordPressLogo, WordPressWordmark } from '../index';
```
Comment (tyxla) [inline packages/components/src/wordpress-logo/docs/example.jsx:1]: Same as above with regards to moving to another PR.

File: packages/components/src/wordpress-logo/index.tsx [hunk 7]
```diff
@@ -1,19 +1,46 @@
...
 		</svg>
 	);
 };
 
-export default WordPressLogo;
+interface WatermarkProps extends Omit< LogoProps, 'size' > {
+	size?: { width: number; height: number };
+}
+
+export const WordPressWordmark: React.FunctionComponent< WatermarkProps > = ( {
```
Comment (tyxla) [inline packages/components/src/wordpress-logo/index.tsx:28]: Why isn't the `WordPressWordmark` in its own separate file? In an ideal world, we may want to bundle only one or the other and not add extra unused bytes.

File: packages/wpcom-template-parts/package.json [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+	"bugs": "https://github.com/Automattic/wp-calypso/issues",
+	"types": "dist/types",
+	"scripts": {
+		"clean": "tsc --build ./tsconfig.json ./tsconfig-cjs.json --clean && rm -rf dist",
+		"build": "tsc --build ./tsconfig.json ./tsconfig-cjs.json",
+		"prepack": "yarn run clean && yarn run build",
+		"watch": "tsc --build ./tsconfig.json --watch"
+	},
+	"dependencies": {
+		"@automattic/components": "workspace:^",
```
Comment (tyxla) [inline packages/wpcom-template-parts/package.json:33]: Let's make sure we're including all dependencies. For example, I see we're using `i18n -calypso` and it's not specifically listed here.

File: packages/wpcom-template-parts/src/types.d.ts [hunk 9]
```diff
@@ -0,0 +1,12 @@
+declare module '@automattic/components' {
+	type Props = {
+		className?: string;
+		size?: { width: number; height: number };
+	};
+
+	export const WordPressWordmark: React.FunctionComponent< Props >;
+}
```
Comment (tyxla) [inline packages/wpcom-template-parts/src/types.d.ts:8]: Is this something we can declare in the `@automattic/components` package itself?

File: packages/wpcom-template-parts/src/types.ts [hunk 10]
```diff
@@ -0,0 +1,23 @@
+export interface HeaderProps {
+	isLoggedIn: boolean;
+	sectionName: string;
+}
+
+export interface FooterProps {
+	isLoggedIn: boolean;
+	currentRoute: string;
+	additonalFooterLinks: any;
```
Comment (tyxla) [inline packages/wpcom-template-parts/src/types.ts:9]: Any more specific type we can use here?

File: packages/wpcom-template-parts/src/universal-footer-navigation/index.tsx [hunk 11]
```diff
@@ -445,14 +430,30 @@ const UniversalNavbarFooter = () => {
...
+						</svg>
```
Comment (tyxla) [inline packages/wpcom-template-parts/src/universal-footer-navigation/index.tsx:445]: Opportunity for further componentization as we did with the WordPress wordmark?

File: packages/wpcom-template-parts/src/universal-footer-navigation/index.tsx [hunk 12]
```diff
@@ -445,14 +430,30 @@ const UniversalNavbarFooter = () => {
...
+						{ automatticBranding.noun }
+					</a>
+					<div className="lp-logo-label-spacer"></div>
+					<a className="lp-link-work" href="https://automattic.com/work-with-us/">
```
Comment (tyxla) [inline packages/wpcom-template-parts/src/universal-footer-navigation/index.tsx:449]: Are those links something we should localize with `localizeUrl()` for example?

Comment (tyxla) [inline packages/wpcom-template-parts/src/universal-footer-navigation/index.tsx:449]: Are we intentionally staying in the same document/tab and not opening a new one?

**Comment by arcangelini** [comment]
Everyone, please forgive me as this PR was a huge learning experience. There were some errors caused by something I had done in my PR, but I was not able to find them. I spent the past two days following build logs, package dependancies and type errors till last night when I tried to fix the errors that were mentioned above and made an even bigger mess.

This is a lesson learned for me in chunking large work into smaller appropriately scoped PR's. I am abandoning this PR in favor of 3 separate PR's. Please forgive me for pinging everyone and thank you for your grace.

1. Move WordPressWordmark to package - #73583 
2. Create a new package for the header and footer - #73585 
3. Replace the Calypso component - to be created
